app/services/payment_requests/payments/stripe_service.rb

```diff
@@ -82,13 +82,39 @@ def generate_payment_url
           }
         )
 
-        result.payment_url = result_url['url']
+        result.payment_url = result_url["url"]
 
         result
       rescue Stripe::CardError, Stripe::InvalidRequestError, Stripe::AuthenticationError, Stripe::PermissionError => e
         deliver_error_webhook(e)
 
-        result.single_validation_failure!(error_code: 'payment_provider_error')
+        result.single_validation_failure!(error_code: "payment_provider_error")
+      end
+
+      def update_payment_status(organization_id:, provider_payment_id:, status:, metadata: {})
+        # TODO: do we have one-time payments for payment requests?
+        payment = if metadata[:payment_type] == "one-time"
+          create_payment(provider_payment_id:, metadata:)
+        else
+          Payment.find_by(provider_payment_id:)
+        end
+
+        return handle_missing_payment(organization_id, metadata) unless payment
+
+        result.payment = payment
+        result.payable = payment.payable
+        return result if payment.payable.payment_succeeded?
+
+        payment.update!(status:)
+
+        processing = status == "processing"
+        payment_status = payable_payment_status(status)
+        update_payable_payment_status(payment_status:, processing:)
+        update_invoices_payment_status(payment_status:, processing:)
```

Comment:
what about doing it inside a `ActiveRecord::Base.transaction`?

Reply:
I would rather leave as it is... these methods sent webhooks, if wrapping them within a transaction the db change would be rolled back while webhook will be sent to users...